### ISSUE TITLE
[Proposal] Better branching strategy

### ISSUE THREAD

**Issue by nchashch** [amethyst/amethyst#203]
In order to better comply with [semver](http://semver.org/) I propose instead of having one `develop` branch to have a `minor` and a `patch` branches.

Before `1.0.0` non breaking changes would live on `patch` branch and breaking changes would live on `minor` branch. After `1.0.0` we would have to add a `major` branch. Then changes would live on their corresponding branches according to [semver](http://semver.org/).

It would make sense to do this after #202 is resolved.

todo:
- [ ] update CONTRIBUTING.md file
- [ ] add `minor` and `patch` branches
- [ ] remove `develop` branch

**Comment by torkleyy** [comment]
Should the `release-*` branches still exist? What is the advantage over starting non breaking PRs against `release-*` and breaking against `develop`?

**Comment by nchashch** [comment]
AFAICT `release-*` branches are there to allow retractively adding bug-fixes to old releases, not to separate breaking and non breaking changes.

Using `minor` and `patch` branches for separating braking/non breaking changes would be less confusing than using `release-*` and `develop` branches. Also currently nobody uses `release-*` branches this way and both breaking/non breaking changes just go into `develop`.

**Comment by nchashch** [comment]
I guess if we were to go `minor`, `patch` branches way removing `release-*` branches might make sense. Tags should be enough to access older versions.

**Comment by torkleyy** [comment]
OK, I agree with this proposal

**Comment by ebkalderon** [comment]
@nchashch I personally feel like GitFlow may be a better choice, given that it's a more well-known model and is pretty close to our current setup. Additionally, there is a [`git flow` subcommand for Git](https://github.com/nvie/gitflow) available to automate the whole thing, which should simplify the `CONTRIBUTING.md` file a lot if everyone decides to just use that.

**Comment by agmcleod** [comment]
On MelonJS we have typically checked out a new 3.x or 2.x branch when we start making breaking changes from 2.0 on master. So we have a branch we can patch/do non breaking changes to, while doing our breaking changes on master for a 3.0 release. Might be harder to manage with more developers though.

**Comment by nchashch** [comment]
@ebkalderon If you added a new branch for every feature and then merged it directly into `develop` breaking and non breaking changes would still be mixed together. So if one small breaking change was merged into `develop` we wouldn't be able to release a PATCH bump, and all bug fixes/docs improvements would have to wait until the next MINOR bump (which should be rare significant breaking changes) to be released.

Basically I want to separate PATCH level changes and MINOR level changes, so all the non breaking changes can become available as soon as possible with frequent PATCH releases, while work for the next MINOR release can still continue on a separate branch.

**Comment by torkleyy** [comment]
I just read through [this](http://nvie.com/posts/a-successful-git-branching-model/) and I have to say that I'm really convinced now. We should definitly use this GitFlow branching model.

**Comment by torkleyy** [comment]
@nchashch

What about just creating a `patch-add-foo` branch and then merging it into master and develop?

**Comment by vitvakatu** [comment]
I agree with @torkleyy

**Comment by ebkalderon** [comment]
@nchashch Perhaps you are misunderstanding the point of the `release-*` branches vs. `develop`.

1. Bug fix:
   * Developer
       1. Checkout `release-*` branch and work on it.
       2. Submit PR to `release-*` branch.
   * Maintainer
       1. Increment PATCH version number.
       2. Merge to `develop`.
       3. Merge into `master` and tag the new release.
       4. Publish to crates.io.
2. Feature or change (breaking or not)
    * Developer
       1. Branch from `develop` into `feature-*` and work on it.
       2. Submit PR to `develop`.
    * Maintainer
       1. Accept PR, increment MAJOR version number.
       2. Merge into `develop`.
       3. Branch from `develop` into new `release-X.Y` branch.
       4. Merge into `master` and tag the new release.
       5. Publish to crates.io.

Right now we do not distinguish between breaking and non-breaking since any batch of changes deserves a release, and we should ship early and often, although I personally haven't been following that close enough as I should 😛. Two takeaways:

1. We should decide how many feature changes or bugfixes it takes to warrant a new release.
2. We should branch from `develop` when working on features instead of submitting directly to `develop.

Please feel free to correct me if I made any glaring mistakes in my explanation!

**Comment by nchashch** [comment]
@ebkalderon Yes, I see. `release-*` branches are for preparing releases/last minute fixes, not for retroactively fixing things.

I am fine with sticking to GitFlow if we can figure out how to cleanly separate breaking and non breaking changes.

The problem with @torkleyy's  proposal for using `patch-add-foo` style branches and merging them into both `master` and `develop` is that you would have to push to `master` for every non breaking change, which amounts to a PATCH release. There is no single shared branch for all non breaking changes.

Spliting `develop` branch into `minor` and `patch` would allow accumulating non breaking changes in `patch`, and when there is a critical number of them merge it into `master`. We can still keep all the other parts of GitFlow, like `release-*` branches, `feature` branches, `hotfix-*` branches.

**Comment by Xaeroxe** [comment]
Closing due to inactivity.
